packages/flyout/src/getCoordinates.js

```diff
@@ -0,0 +1,262 @@
+    case TOP_CENTER:
+    case BOTTOM_CENTER:
+      return panelRect.width / 2 - CARET_SIZE / 2;
+    case TOP_LEFT:
+    case BOTTOM_LEFT:
+      return actionRect.width / 2 - CARET_SIZE / 2;
+    case TOP_RIGHT:
+    case BOTTOM_RIGHT:
+      return panelRect.width - (actionRect.width / 2 + CARET_SIZE / 2);
+    default:
+      return DEFAULT_COORDINATES.pointerPosition.left;
+  }
+}
+
+/**
+ * @param {Payload} payload
+ * @returns {number}
+ */
+function calculatePanelContainerTopOffset({
+  anchorPoint,
+  actionRect,
+  panelRect
+}) {
+  switch (anchorPoint) {
+    case TOP_LEFT:
+    case TOP_CENTER:
+    case TOP_RIGHT:
+      return actionRect.height + CARET_SIZE / 2;
+    case BOTTOM_LEFT:
+    case BOTTOM_CENTER:
+    case BOTTOM_RIGHT:
+      return -1 * panelRect.height - CARET_SIZE / 2;
+    case LEFT_TOP:
+    case RIGHT_TOP:
+      return 0;
+    case LEFT_CENTER:
+    case RIGHT_CENTER:
+      return (panelRect.height - actionRect.height) / -2;
+    case LEFT_BOTTOM:
+    case RIGHT_BOTTOM:
+      return -1 * (panelRect.height - actionRect.height);
+    default:
+      return DEFAULT_COORDINATES.containerPosition.top;
+  }
+}
+
+/**
+ * @param {Payload} payload
+ * @returns {number}
+ */
+function calculatePanelContainerLeftOffset({
+  anchorPoint,
+  actionRect,
+  panelRect
+}) {
+  switch (anchorPoint) {
+    case LEFT_TOP:
+    case LEFT_CENTER:
+    case LEFT_BOTTOM:
+      return actionRect.width + CARET_SIZE / 2;
+    case RIGHT_TOP:
+    case RIGHT_CENTER:
+    case RIGHT_BOTTOM:
+      return -1 * panelRect.width - CARET_SIZE / 2;
+    case TOP_LEFT:
+    case BOTTOM_LEFT:
+      return 0;
+    case TOP_CENTER:
+    case BOTTOM_CENTER:
+      return (panelRect.width - actionRect.width) / -2;
+    case TOP_RIGHT:
+    case BOTTOM_RIGHT:
+      return -1 * (panelRect.width - actionRect.width);
+    default:
+      return DEFAULT_COORDINATES.containerPosition.top;
+  }
+}
+
+/**
+ * @param {Payload} payload
+ * @returns {Coordinates}
+ */
+function calculatePointerPosition(props) {
+  return {
+    top: calculatePointerTopOffset(props),
+    left: calculatePointerLeftOffset(props)
+  };
+}
+
+/**
+ * @param {Payload} payload
+ * @returns {Coordinates}
+ */
+function calculatePanelContainerPosition(props) {
+  return {
+    top: calculatePanelContainerTopOffset(props),
+    left: calculatePanelContainerLeftOffset(props)
+  };
+}
+
+/**
+ * @param {Payload} payload
+ * @returns {Coordinates}
+ */
+function calculateCoordinates(props) {
+  return {
+    anchorPoint: props.anchorPoint,
+    containerPosition: calculatePanelContainerPosition(props),
+    pointerPosition: calculatePointerPosition(props)
+  };
+}
+
+/**
+ * Determines whether the given position is entirely within the viewport
+ * @param {Payload} payload
+ * @returns {function(Coordinates): boolean}
+ */
+function createViewportDeterminer(props) {
+  const { viewportRect, panelRect, actionRect } = props;
+
+  return function isInViewport({ containerPosition }) {
```

Comment:
What is the advantage of returning a function here, instead of having a method `isInViewport` ?

Reply:
It's mainly so that I can pass it directly into `Array.prototype.find`. I'll refactor since it's mainly my preference.